### PR TITLE
add a security module （#628）

### PR DESCRIPTION
**Pull request by aqa510415008**
[#628](https://github.com/ThreeMammals/Ocelot/issues/628)

Add security mechanisms, IP whitelists, blacklists, and extended security interfaces.
Extend security by implementing ISecurityPolicy。

@TomPallister Evaluation is feasible, feasible, I add unit test

### CODE REVIEW

File: src/Ocelot/Configuration/File/FileSecurityOptions.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class FileSecurityOptions
+    {
+        public FileSecurityOptions()
+        {
+            IPWhitelist = new List<string>();
+            IPBlacklist = new List<string>();
+        }
+
+        public List<string> IPWhitelist { get; set; }
```
Comment (TomPallister) [inline src/Ocelot/Configuration/File/FileSecurityOptions.cs]: Call this IPAllowedList

File: src/Ocelot/Configuration/File/FileSecurityOptions.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public FileSecurityOptions()
+        {
+            IPWhitelist = new List<string>();
+            IPBlacklist = new List<string>();
+        }
+
+        public List<string> IPWhitelist { get; set; } 
+
+        public List<string> IPBlacklist { get; set; }
```
Comment (TomPallister) [inline src/Ocelot/Configuration/File/FileSecurityOptions.cs]: Call this IPBlockedList

File: src/Ocelot/Configuration/SecurityOptions.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class SecurityOptions
+    {
+        public SecurityOptions(List<string> whitelist,List<string> blacklist)
+        {
+            this.IPBlacklist = blacklist;
+            this.IPWhitelist = whitelist;
+        }
+
+        public List<string> IPWhitelist { get; private set; }
```
Comment (TomPallister) [inline src/Ocelot/Configuration/SecurityOptions.cs]: Call this IPAllowedList

File: src/Ocelot/Configuration/SecurityOptions.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+        public List<string> IPBlacklist { get; private set; }
```
Comment (TomPallister) [inline src/Ocelot/Configuration/SecurityOptions.cs]: Call this IPBlockedList

File: src/Ocelot/Security/IPSecurity/IPSecurityPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            SecurityOptions securityOptions = context.DownstreamReRoute.SecurityOptions;
+            if (securityOptions == null)
+            {
+                return new OkResponse();
+            }
+            if (securityOptions.IPBlacklist != null)
+            {
+                if (securityOptions.IPBlacklist.Exists(f => f == clientIp.ToString()))
+                {
+                    var error = new UnauthenticatedError($"{clientIp.ToString()} Cannot request to enter the blacklist");
```
Comment (TomPallister) [inline src/Ocelot/Security/IPSecurity/IPSecurityPolicy.cs]: blockedlist instead of blacklist

**Comment by aqa510415008** [comment]
Security protection extension for Ocelot AuthorizationToken

[Ocelot.Security.AuthorizationToken.EntityFramework](https://github.com/aqa510415008/Ocelot.Security.AuthorizationToken.EntityFramework)

**Comment by TomPallister** [comment]
> Security protection extension for Ocelot AuthorizationToken
> 
> [Ocelot.Security.AuthorizationToken.EntityFramework](https://github.com/aqa510415008/Ocelot.Security.AuthorizationToken.EntityFramework)

very cool :D

**Comment by TomPallister** [comment]
@aqa510415008 thank you for the tests, this is very good. I have merged it now.

**Comment by TomPallister** [comment]
@aqa510415008 thank you for the tests, this is very good. I have merged it now.

**Pull request merged**